### ISSUE TITLE
[BUG] StatePrep adjoint does not properly uncompute state

### ISSUE THREAD

**Issue by anthonysmaldone** [PennyLaneAI/pennylane#5099]
### Expected behavior

I expected that when I apply `StatePrep` to the ground state to prepare some arbitrary state, that `qml.adjoint(qml.StatePrep))` will return me back to ground state with a phase angle of 0. 

### Actual behavior

The adjoint of `StatePrep` is off by a global.

### Additional information

I need to perform the adjoint of StatePrep while being controlled by an ancilla qubit, so the global phase offset manifests into a problem. Manually adding a phase gate that corrects the phase with respect to the number of negative amplitudes in original state works, but leaves the circuit un-batchable.

Functional example that does not allow batching:

```
import pennylane as qml

state = np.array([-0.5, 0.2, 0.3, 0.9, 0.5, 0.2, 0.3, 0.9], requires_grad=True)
state = state / np.linalg.norm(state)

test = qml.device('default.qubit', wires=3)
@qml.qnode(test)
def test_circ(state):
    qml.StatePrep(state=state,wires=[0,1,2])
    qml.adjoint(qml.StatePrep(state=state,wires=[0,1,2]))
    #count the number of negative amplitudes in the initial state
    #to know how to adjust the global phase
    ang = np.sum(state < 0)/(state.shape[-1]) * np.pi
    
    qml.RZ(phi=ang*2)
    #qml.GlobalPhase(phi=ang)

    return qml.state()
```

### Source code

```shell
import pennylane as qml

state = np.array([-0.5, 0.2, 0.3, 0.9, 0.5, 0.2, 0.3, 0.9], requires_grad=True)
state = state / np.linalg.norm(state)

test = qml.device('default.qubit', wires=3)

@qml.qnode(test)
def test_circ(state):
    qml.StatePrep(state=state,wires=[0,1,2])
    qml.adjoint(qml.StatePrep(state=state,wires=[0,1,2]))
    # optionally apply some RZ shift to correct the phase offset
    # in this specific case, it is off by pi/8
    # qml.RZ(phi=np.pi/4,wires=0)

    return qml.state()

print(test_circ(state))
```

### Tracebacks

```shell
Output:
[ 9.23879533e-01+3.82683432e-01j -9.68859404e-18+5.12040449e-17j
  1.96261557e-16+1.27570012e-16j  1.70680150e-17+2.90657821e-17j
 -1.11022302e-16+0.00000000e+00j  9.68859404e-18-9.68859404e-18j
  3.92523115e-17-9.81307787e-18j -1.70680150e-17+1.70680150e-17j]
```

### System information

```shell
Name: PennyLane
Version: 0.34.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: PennyLane-Lightning

Platform info:           Linux-6.1.58+-x86_64-with-glibc2.35
Python version:          3.10.12
Numpy version:           1.23.5
Scipy version:           1.11.4
Installed devices:
- default.gaussian (PennyLane-0.34.0)
- default.mixed (PennyLane-0.34.0)
- default.qubit (PennyLane-0.34.0)
- default.qubit.autograd (PennyLane-0.34.0)
- default.qubit.jax (PennyLane-0.34.0)
- default.qubit.legacy (PennyLane-0.34.0)
- default.qubit.tf (PennyLane-0.34.0)
- default.qubit.torch (PennyLane-0.34.0)
- default.qutrit (PennyLane-0.34.0)
- null.qubit (PennyLane-0.34.0)
- lightning.qubit (PennyLane-Lightning-0.34.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by timmysilv** [comment]
Hi @anthonysmaldone, thanks for bringing this up! A few quick things to note:
1. State-prep operations are decomposed when applied mid-circuit (plus, the adjoint of a state-prep op is not considered a state-prep op anyway), so the second operation gets decomposed using the Mottonen decomposition which I believe would introduce this phase
2. There is certainly the option to consider adding a global phase to the decomposition, but I'm not entirely sure if this will be a straightforward thing to do given how we handle state-prep operations
3. State-prep operations used mid-circuit are stated to not act as an arbitrary unitary that takes any state to any other state, but rather one that meets the requirement $U|0\rangle = |\psi\rangle$ as described in [the operator's doc](https://docs.pennylane.ai/en/stable/code/api/pennylane.StatePrep.html)

**Comment by timmysilv** [comment]
small follow-up: as indicated in its docs, `StatePrep` decomposes to [MottonenStatePrep](https://docs.pennylane.ai/en/stable/code/api/pennylane.MottonenStatePreparation.html), and that is only correct up to a global phase. So my third point isn't quite right, in the sense that we actually land up with $e^{i \phi} |\psi\rangle$. I would categorize this as semi-expected behaviour, but not exceedingly obvious behaviour. The original author of the Möttönen decomposition chose to ignore the global phase, but given that we now support them in PennyLane, we can look into adding it to the decomposition. nice find!

**Comment by astralcai** [comment]
Can we close this issue now?

**Comment by timmysilv** [comment]
yep! fixed in #5166
